### ISSUE TITLE
Disposing FlurlClient disposes HttpClient affecting other threads

### ISSUE THREAD

**Issue by rcollette** [tmenier/Flurl#308]
We have a ASP.net core transient service, using a singleton instance of `PerBaseUrlClientFactory` via constructor DI.  The transient service creates a FlurlClient during construction from the injected factory.  The transient service implements IDisposable and calls dispose on FlurlClient.

The problem seems to be that when the FlurlClient is disposed, is is also disposing the HttpClient it contains, which gets shared among multiple/simultaneous requests.  The disposal winds up causing an error like:
```
System.AggregateException: One or more errors occurred. (GET http://host/path timed out.) ---> Flurl.Http.FlurlHttpTimeoutException: GET http://host/path timed out. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.

   at async Task<HttpResponseMessage> System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task<HttpResponseMessage> sendTask, HttpRequestMessage request, CancellationTokenSource cts, bool disposeCts)

   at async Task<HttpResponseMessage> Flurl.Http.FlurlRequest.SendAsync(HttpMethod verb, HttpContent content, Nullable<CancellationToken> cancellationToken, HttpCompletionOption completionOption)
```

Wouldn't you want to make the factory implement IDisposable and have the HttpClient instances disposed when the factory is disposed during shutdown?   I think there is a general design principle "That which creates, contains. That which creates, destroys." or something of that sort, that might apply here.

**Comment by rcollette** [comment]
Or going old school, the Microsoft Data Access block used to use a counter to determine if in fact the last connection.close was called before actually closing a shared dbConnection.  Something similar might be done with the disposal.   So increment a counter when handing out a shared instance and decrement when dispose is called and only actually dispose when the counter goes back to zero.

**Comment by tmenier** [comment]
A `FlurlClient`'s lifetime is tightly bound to the `HttpClient` it wraps (except for the fact that creation the latter is usually deferred until it's actually needed). The fact that disposing the former disposes the latter is by design. Why are you disposing `FlurlClient` while it's still being used?

**Comment by rcollette** [comment]
Request 1 Creates FlurlClient in the service. at t1
At t1+N Request 2  Creates a new instance of the transient service and thus another FlurlClient but they share the same underlying HttpClient handed out by the factory.

At t1+N+N2 Request1 completes, but Request2 is in process and Service instance 1 disposes and in turn disposes its FlurlClient and the underlying shared instance of HttpClient before request 2 has completed.

Request 1/ServiceInstance1 isn't aware of Request 2/ServiceInstance2, it can't wait on disposal.

**Comment by rcollette** [comment]
OK.  Understood.   

However, I'm not seeing where in Flurl, HttpClient or FlurlClient is disposed.   The FlurlClient isn't instantiated by the DI framework so, that won't dispose of it.  I don't see Dispose being called on anything but the request nor any `using` statements for HttpClient or FlurlClient.

Usually anything creating a disposable would be responsible for disposing it, in a `using` statement for example at a method level or in a Dispose method implementation of IDisposable if the object is retained as, or in, a class member field.

GC doesn't call Dispose, so if I don't what will?

**Comment by tmenier** [comment]
There's nothing in the library that disposes FlurlClient because the library can't possibly know when you're done with it. It's a little like saying nothing in System.Net.Http disposes HttpClient.

That said, I think what you're saying is the factory created it, so the factory should dispose it. I could see wanting a factory-level method that disposes/clears all FlurlClient instance it contains, so I might be open to having that implement IDisposable. I don't think there would be any need for reference counters or anything though - you (or the IoC container) would just call `Dispose` on the factory _instead_ if the individual clients. Would that be a satisfactory solution?

It should be noted too that [disposing HttpClient isn't very important](https://stackoverflow.com/questions/15705092/do-httpclient-and-httpclienthandler-have-to-be-disposed), so neither is disposing FlurlClient. In most cases you want to keep those instances alive for the lifetime of your application, in which case you don't need to dispose them at all. But in cases where your application continues to run after you know you won't be needing those instances anymore, it could be useful.

**Comment by rcollette** [comment]
The only issue I can think of is that a TCP reset/rst might happen when the application ends rather than a TCP close, if there are active requests.  I've seen this when not disposing the .NET SMTP client and that led to some issue that I cannot recall at the moment.

Assuming there are no negative side effects of not disposing HttpClient when shutting down the application, then I guess this isn't necessary.  I'm going to assume that there are no negative affects otherwise you probably would have heard of them.   

In my particular use case, the factory lasts for the lifetime of the application and even on application restart we are more than likely doing a docker container restart as well.  On my side, there probably isn't an impact.   

I think we've batted this around enough.   You have far more experience here.  I'm changing my code to not dispose at all and assume based on your experience that there is no issue with this.

From my point of view this can be closed unless you want to ponder it more.

**Comment by tmenier** [comment]
I think you convinced me something could be done here more than you think. :)

I agree that in your case you shouldn't need to worry about not explicitly disposing those instances when the application shuts down. As I understand it, finalizers _should_ still get hit, and that's the important part. But I do think as creators/cachers of FlurlClients, making factories disposable makes sense. I created a new issue for that (#309). Thanks for the suggestion.
